File: android/src/main/kotlin/com/example/plugin_using_protobuf/utils/AbstractSubscribeModel.kt
package com.example.plugin_using_protobuf.utils
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.fossil.wear3_flutter_plugin.wear3_flutter_plugin.utils.ChannelUtil
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.*

abstract class AbstractSubscribeModel<M, A>(
    val binaryMessenger: BinaryMessenger,
    val methodChannelId: String = UUID.randomUUID().toString(),
    val eventChannelId: String = UUID.randomUUID().toString(),
) : MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler where M : com.google.protobuf.MessageLite, A : com.google.protobuf.MessageLite {
    init {
        this.setup()
    }

    private var isListening: Boolean = false

    private lateinit var _methodChannel: MethodChannel

    private lateinit var _eventChannel: EventChannel
    private var _eventSink: EventChannel.EventSink? = null

    abstract fun buildAction(data: ByteArray): A

    abstract fun handleAction(action: A, result: MethodChannel.Result)

    abstract fun onStartListening()

    abstract fun onStopListening()

    // method channel
    override fun onMethodCall(method: MethodCall, result: MethodChannel.Result) {
        // call action
        if (method.method == "action") {
            val actionData = method.arguments as HashMap<*, *>
            val data = actionData["data"] as ByteArray
            val actionModel = this.buildAction(data)
            this.handleAction(actionModel, result)
        } else if (method.method == "stopListening") {
            isListening = false
            this.onStopListening()
        } else if (method.method == "startListening") {
            if (!isListening) {
                this.onStartListening()
                isListening = true
            }
        } else if (method.method == "dispose") {
            result.success(null)
            this.dispose()
        }
    }

    //Event channel
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this._eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        Log.d("Wear3", "[WEAR3] onDispose cancelEventChannel ${this.javaClass.name}")
        this._eventChannel.setStreamHandler(null)
        this._eventSink = null
        this.onStopListening();
        isListening = false
    }


    private fun setup() {
        Log.d("AbstractSubscribeModel", "Start setup ${this::class.java.name}")
        this._methodChannel = ChannelUtil.INSTANCE.createMethodChannel(methodChannelId, binaryMessenger)
        this._eventChannel = ChannelUtil.INSTANCE.createEventChannel(eventChannelId, binaryMessenger)
        this._methodChannel.setMethodCallHandler(this)
        this._eventChannel.setStreamHandler(this)
        Log.d("AbstractSubscribeModel", "End setup ${this::class.java.name}")
    }

    fun sink(message: M) {
        Handler(Looper.getMainLooper()).post {
            this._eventSink?.success(mapOf("data" to message.toByteArray()))
        }
    }

    open fun dispose() {
        this._methodChannel.setMethodCallHandler(null)
    }
}

File: ios/Classes/utils/AbstractSubscribeModel.swift
import Flutter
import Foundation
import SwiftProtobuf

class AbstractSubscribeModel<Result: SwiftProtobuf.Message, Action: SwiftProtobuf.Message>: NSObject, FlutterStreamHandler {
    private var _eventChannel: FlutterEventChannel?
    private var _methodChannel: FlutterMethodChannel?
    private var _eventSink: FlutterEventSink?
    public var methodChannelId: String
    public var eventChannelId: String
    private var isListening = false
    
    func onStartListening() {
        preconditionFailure("This method must be overridden")
    }
    
    func onStopListening() {
        preconditionFailure("This method must be overridden")
    }

    func buildAction(data: Data) -> Action? {
        do {
            return try Action(serializedData: data)
        } catch {
            print("buildAction error: \(error)")
        }
        return nil
    }

    func handleAction(action _: Action, result: @escaping FlutterResult) {
        preconditionFailure("This method must be overridden")
    }

    func sink(result: Result) {
        do {
            if(_eventSink != nil) {
                print("[iOS] [\(String(describing: type(of: self)))] sink: \(try! result.jsonString())")
                try _eventSink!(["data": result.serializedData(partial: false)])
            }
        } catch {}
    }

    func onListen(withArguments _: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        print("[iOS] [\(String(describing: type(of: self)))] onListen")
        _eventSink = events
        return nil
    }

    func onCancel(withArguments _: Any?) -> FlutterError? {
        print("[WEAR3] onDispose cancelEventChannel [iOS] [\(String(describing: type(of: self)))] onCancel")
        self._eventChannel?.setStreamHandler(nil)
        self._eventSink = nil
        self.onStopListening()
        self.isListening = false
        
        ChannelUtil.shared.cleanUpEventChannel(for: eventChannelId)
        ChannelUtil.shared.cleanUpStreamHandler(for: eventChannelId)
        return nil
    }
    
    func dispose() {
        print("[iOS] [\(String(describing: type(of: self)))] dispose")
        self._methodChannel?.setMethodCallHandler(nil)
        ChannelUtil.shared.cleanUpMethodChannel(for: methodChannelId)
    }

    init(methodChannelId: String? = nil, eventChannelId: String? = nil) {
        self.methodChannelId = methodChannelId ?? UUID().uuidString
        self.eventChannelId = eventChannelId ?? UUID().uuidString
        _methodChannel = ChannelUtil.shared.createMethodChannel(id: self.methodChannelId)
        _eventChannel = ChannelUtil.shared.createEventChannel(id: self.eventChannelId)
        super.init()
        
        
        _methodChannel?.setMethodCallHandler { [weak self] call, result in
            if call.method == "action" {
                let args = call.arguments as! [String:Any]
                let actionData = args["data"] as! FlutterStandardTypedData
                let actionModel = self?.buildAction(data: actionData.data)
                print("[iOS] [\(String(describing: type(of: self)))] action: \(try! actionModel!.jsonString())")
                if actionModel != nil {
                    self?.handleAction(action: actionModel!, result: result)
                }
            } else if (call.method == "dispose") {
                result(nil)
                self?.dispose()
            } else if (call.method == "stopListening") {
                self?.isListening = false
                self?.onStopListening()
                result(nil)
            } else if(call.method == "startListening") {
                if(self?.isListening == false) {
                    self?.onStartListening()
                    self?.isListening = true
                }
                result(nil)
            }
        }
        
        ChannelUtil.shared.registerStreamHandler(id: self.eventChannelId, handler: self)
        _eventChannel?.setStreamHandler(self)
    }
}
